bindings/rust/aws-kms-tls-auth/src/provider.rs

```diff
@@ -0,0 +1,222 @@
+    ///
+    /// KMS ciphertexts have observable regularities in their structure. Obfuscating
+    /// the identity prevents any of that from being observable over the wire.
+    obfuscation_key: Arc<ObfuscationKey>,
+    /// The current Psk being set on all new connections.
+    ///
+    /// The lock is necessary because this is updated every 24 hours by the
+    /// background updater.
+    psk: Arc<RwLock<s2n_tls::psk::Psk>>,
+    /// The last time the key was updated. If `None`, then a key update is in progress.
+    last_update: Arc<RwLock<Option<Instant>>>,
+}
+
+impl PskProvider {
+    /// Initialize a `PskProvider`.
+    ///
+    /// This method will call the KMS generate-data-key API to create the initial
+    /// PSK that will be used for TLS connections.
+    pub async fn initialize(
+        client: Client,
+        key: KeyArn,
+        obfuscation_key: ObfuscationKey,
+    ) -> anyhow::Result<Self> {
+        let psk = Self::generate_psk(&client, &key, &obfuscation_key).await?;
+
+        let value = Self {
+            client: client.clone(),
+            kms_key_arn: Arc::new(key),
+            obfuscation_key: Arc::new(obfuscation_key),
+            psk: Arc::new(RwLock::new(psk)),
+            last_update: Arc::new(RwLock::new(Some(Instant::now()))),
+        };
+        Ok(value)
+    }
+
+    /// Check if a key update is needed. If it is, kick off a background task
+    /// to call KMS and create a new PSK.
+    fn maybe_trigger_key_update(&self) {
+        let last_update = match *self.last_update.read().unwrap() {
+            Some(update) => update,
+            None => {
+                // update already in progress
+                return;
+            }
+        };
+
+        if last_update.elapsed() > KEY_ROTATION_PERIOD {
+            // because we released the lock above, we need to recheck the update
+            // status after acquiring the lock.
+            let mut reacquired_update = self.last_update.write().unwrap();
+            if reacquired_update.is_some() {
+                *reacquired_update = None;
+                tokio::spawn({
+                    let psk_provider = self.clone();
+                    async move {
+                        psk_provider.rotate_key(last_update).await;
+                    }
+                });
+            }
+        }
+    }
+
+    pub async fn rotate_key(&self, previous_update: Instant) {
+        match Self::generate_psk(&self.client, &self.kms_key_arn, &self.obfuscation_key).await {
+            Ok(psk) => {
+                *self.psk.write().unwrap() = psk;
+                *self.last_update.write().unwrap() = Some(Instant::now());
+            }
+            Err(e) => {
+                // we failed to update the PSK. Restore the previous update and let
+                // someone else try.
+                tracing::error!("failed to create PSK from KMS {e}");
+                *self.last_update.write().unwrap() = Some(previous_update);
+            }
```

Comment:
Can you walk me through the failure case? If we fail to rotate the key, how frequently / often will we retry?